File: leetcode/binary_search/69(easy)-Sqrt.java
public class Sqrt {
  public int mySqrt(int x) {
    long left = 0;
    long right = x;
    long mid = 0;

    while(left <= right) {
      mid = (left + right) / 2;
      long sqrt = mid * mid;

      if (x == sqrt) {
        return (int)mid;
      }

      if (x < sqrt) {
        right = mid - 1;
      }

      if (x > sqrt) {
        left = mid + 1;
      }
    }

    if (left * left < x) {
      return (int)left;
    }

    return (int)right;
  }
}

File: leetcode/binary_search/69(easy) Sqrt.js
/**
 * @param {number} x
 * @return {number}
 */
var mySqrt = function (x) {
  if (x === 1 || x === 0) {
    return x
  }
  let start = 0
  let end = x

  let ans = start
  while (start <= end) {
    const mid = Math.floor((start + end) / 2)

    if (mid * mid === x) {
      ans = mid
      break
    }

    if (mid * mid < x) {
      ans = mid // 최대값 저장
      start = mid + 1
    }

    if (mid * mid > x) {
      end = mid - 1
    }
  }

  return ans
}
